astropy/units/tests/test_format.py

```diff
@@ -647,17 +647,17 @@ def test_deprecated_did_you_mean_units():
 @pytest.mark.parametrize("string", ["mag(ct/s)", "dB(mW)", "dex(cm s**-2)"])
 def test_fits_function(string):
     # Function units cannot be written, so ensure they're not parsed either.
+    print(string)
     with pytest.raises(ValueError):
-        print(string)
         u_format.Fits().parse(string)
 
 
 @pytest.mark.parametrize("string", ["mag(ct/s)", "dB(mW)", "dex(cm s**-2)"])
 def test_vounit_function(string):
     # Function units cannot be written, so ensure they're not parsed either.
+    # ct, dex also raise warnings - irrelevant here.
+    warnings.simplefilter("ignore")
     with pytest.raises(ValueError), warnings.catch_warnings():
-        # ct, dex also raise warnings - irrelevant here.
-        warnings.simplefilter("ignore")
```

Comment:
This shouldn't have been moved. If ruff did this, then this should be reported upstream.

Reply:
PT012 is not among rules Ruff can fix automatically.

Reply:
I moved it. Not sure the best way to satisfy PT012 and keep it ignoring warnings. Right now I can think of three options. 
1. Add pytest.mark.filterwarning("ignore") to the function definition. 
2.   Split the context managers out and have the with pytest.raises be the inside one. 
3.  Add a helper function which contains the lines below.

Any other options?

Reply:
I like 1 and 4 (add a noqa).

Reply:
Why can't you just have nested `with`-blocks?

Reply:
Is this about moving the `print(string)` (that seems sensible, though I don't understand the switch to using `__repr__` as a method here – and am not sure if it is necessary at all, as a failed test should give you at least its index),
or about the noqa for the `with`s?
In https://github.com/astropy/astropy/pull/16158/files#r1530953138 it seems to object to the nested blocks, but is it vetoing the single-line version here as well, or to the `simplefilter` inside? In fact the 3 compound forms are supposedly [semantically equivalent](https://docs.python.org/3/reference/compound_stmts.html#the-with-statement), so I don't even see why the argument about misleading raises holds here.

Reply:
Yes it complains about the `warnings.simplefilter("ignore")`. Well more that there are two lines under the raises statement.
 
Nested with blocks like this?
```
with pytest.raises(ValueError):
        with warnings.catch_warnings():
            # ct, dex also raise warnings - irrelevant here.
            warnings.simplefilter("ignore")
            u_format.VOUnit().parse(string)
```
If so I tried that and ruff didn't like it.

Reply:
Doesn't this work if you change the order of the `with`-statements?

Reply:
As long as it is inside `catch_warnings` - otherwise the filter does not get reset.

Reply:
Yes, I think this way around
```python
with warnings.catch_warnings():
    # ct, dex also raise warnings - irrelevant here.
    warnings.simplefilter("ignore")
    with pytest.raises(ValueError):
        u_format.VOUnit().parse(string)
```
it should work correctly and hopefully be accepted by Ruff.